core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -236,13 +267,25 @@ class HadoopMapReduceCommitProtocol(
     committer.setupTask(taskContext)
     addedAbsPathFiles = mutable.Map[String, String]()
     partitionPaths = mutable.Set[String]()
+    dynamicStagingTaskFiles = mutable.Set[Path]()
   }
 
   override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
     val attemptId = taskContext.getTaskAttemptID
     logTrace(s"Commit task ${attemptId}")
     SparkHadoopMapRedUtil.commitTask(
       committer, taskContext, attemptId.getJobID.getId, attemptId.getTaskID.getId)
+    if (dynamicPartitionOverwrite) {
+      val fs = stagingDir.getFileSystem(taskContext.getConfiguration)
+      dynamicStagingTaskFiles.foreach { stagingTaskFile =>
+        val fileName = stagingTaskFile.getName
+        val partitionPath = getDynamicPartitionPath(stagingTaskFile, taskContext)
+        val finalFile = new Path(partitionPath, fileName)
+        if (!fs.exists(finalFile) && !fs.rename(stagingTaskFile, finalFile)) {
```

Comment:
what happens should the `finalFile` exist ?

Reply:
I think the finalFile may exist when spark.hadoop.outputCommitCoordination.enabled is false.
It is true by defaults.

Reply:
It was a dumb question by me. It's by design to skip `rename` when `finalFile` exists to avoid conflicts. As to when the `finalFile` will exist, are you saying there will be no conflict with task speculation if output commit coordination is enabled ? 

Another question is if it's possible the existing `finalFile` is corrupted due to another task crashing during commit.

Reply:
Thanks, I just check the code.
It seems that the  `SparkHadoopMapRedUtil.commitTask` has no meaning for dynamic partition overwrite, because it has a specific staging dir (`.spark-staging-jobId`) instead of use the working temporary of FileOutputCommitter(`_temporary/0`).

So, the conflict issue is not related with the enable of outputCommitCoordination.


> Another question is if it's possible the existing finalFile is corrupted due to another task crashing during commit.

It is possible, because there may be several dynamicStagingTaskFiles，if the task aborted when commitTask, partial outputs are moved to the partition paths.

Reply:
I think `outputCommitCoordination` still avoids two attempts committing at the same time through driver. 

> It is possible, because there may be several dynamicStagingTaskFiles，if the task aborted when commitTask, partial outputs are moved to the partition paths.

I mean what will happen if `rename` fails.

Reply:
If rename failed, partial result would be loss.

Reply:
Do we tests for the two failure scenarios ?

1. task aborted between renaming staging files
2. `rename` fails and an `IOException` is thrown

Reply:
Will try to add relative tests